dev/system-test/firestore.ts

```diff
@@ -2137,6 +2137,22 @@ describe('WriteBatch class', () => {
       });
   });
 
+  it('set supports partials', async () => {
+    const ref = randomCol.doc('doc').withConverter(postConverter);
+    await ref.set(new Post('walnut', 'author'));
+    const batch = firestore.batch();
+    batch.set(ref, {title: 'olive'} as Partial<Post>, {merge: true});
```

Comment:
We need to write our API to ensure that users don't have to do this cast:

```
set<T>(
    documentRef: DocumentReference<T>,
    data: Partial<T>,
    options: { merge: .. } | { mergePaths : ... }
  ): Transaction;
  set<T>(
    documentRef: DocumentReference<T>,
    data: T
  ): Transaction;
```

Reply:
Can't specify all the options since the `SetOptions` type allows both `merge` and `mergeFields` to be passed in, but I made `SetOption` required, which still fixed the need for cast.